squad/core/models.py

```diff
@@ -979,6 +979,42 @@ def full_name(self):
             suite = self.metadata.suite
         return join_name(suite, self.name)
 
+    class Confidence(object):
+        def __init__(self, passes, count, threshold):
+            self.passes = passes
+            self.count = count
+            self.threshold = threshold
+
+        @property
+        def score(self):
+            return 100 * (self.passes / self.count)
+
+    __confidence__ = None
+
+    @property
+    def confidence(self):
+        if self.__confidence__:
+            return self.__confidence__
+
+        date = self.test_run.build.datetime
+        limit = self.build.project.build_confidence_count
+        tests = Test.objects.filter(
+            suite=self.suite,
+            metadata__name=self.name,
+            test_run__build__datetime__lt=date,
+            test_run__environment=self.test_run.environment,
+        ).order_by("-test_run__build__datetime")[:limit]
```

Comment:
making this confidence score like this will indeed cause 1 extra query for each (test,environment) pair in the build. Given how large the tests table is, this will probably cause performance issues. What if you think of another design:

- call your `Confidence` class `FailureWithConfidence` instead
- have a single method in `Build`, called `failures_with_confidence`, that fetches all relevant `Test` objects in a single query, and then create the `FailureWithConfidence` objects interactively.

This might require you to change the design of `FailureWithConfidence` (prev. `Confidence`) slightly, like this:

```python
class FailureWithConfidence(object):
    def __init__(self, passes=0, count=0):
        self.passes = passes
        self.count = count

    def add_pass(self):
        self.passes += 1
        self.count += 1

    def add_failure(self):
        self.count += 1
```

(thread the above as pseudocode, it's untested)

The implementation of `Build.failures_with_confidence` could then look like the body of your view: iterate over the `Test` objects, collect the failure objects in a 2-dimensional array, creating new ones the first time it seems that (test name, environment) pair, and calling either `add_pass` or `add_failure`. You might create Failure objects for all tests, and the later discard the ones where `passes == count`.

Reply:
Question: Is the performance of the patch OK on its own? Or do you think we really do need to change things up?

I'm not sure I follow 100%, but I created a patch that maybe follows along with this. See below.

If the patch is not what you had in mind, do you really think we need to pull all the tests (including the history) at once instead of over multiple queries?

```
... in squad/core/models.py, class Build ...

+  566     __failures__ = None
+  567
+  568     @property
+  569     def failures(self):
+  570         if self.__failures__:
+  571             return self.__failures__
+  572
+  573         failing = self.tests.filter(
+  574             result=False,
+  575             has_known_issues=False,
+  576         )
+  577
+  578         for t in failing:
+  579             limit = self.project.build_confidence_count
+  580             history = Test.objects.filter(
+  581                 metadata=t.metadata,
+  582                 environment=t.environment,
+  583                 build_id__lt=self.id,
+  584             ).order_by("-build_id")[:limit]
+  585
+  586             // t.set_confidence(history)
+  587
+  588         self.__failures = failing
+  589         return self.__failures__
```

Reply:
multiple queries can be ok when you know how many of them you need. in this case, we can easily have hundreds or thousands of them. Think a particularly broken AOSP build where 1% of its 1M tests fail. Then you have 1000 queries.

In this case you can solve it with 2 queries: 1 for "which tests fail on the current build", and 1 for "get the history of this set of tests for the latest N builds", and then you can do the confidence calculation on the python side. The 2nd query is similar to the one you have inside the loop, but with `metadata__in=[t.metadata for t in failing]` instead of `metadata=t.metadata`. Make sure to augment the first query with `prefetch_related("metadata")` as well to avoid have the same 100 queries but against the metadata table.